tls/src/client.rs

```diff
@@ -206,14 +247,90 @@ impl Client {
+
+        // get Nonce from server
+        let server_nonce = recv_msg(&mut stream).await?;
+        let server_nonce = Nonce::try_from(server_nonce)?;
+
+        // get attestation & verify it before sending it
+        // The attesation protocol has an inherent race condition between
+        // getting the log and the attestation. We verify our own attestation
+        // before sending it to the challenger to fail as early as possible.
+        let attest_data = get_attest_data(&attest_config, &server_nonce)?;
+        dice_verifier::verify_attestation(
+            &attest_data.certs[0],
+            &attest_data.attestation,
+            &attest_data.log,
+            &server_nonce,
+        )?;
+
+        // send client attestation cert chain to server
+        let cert_chain_der = certs_to_der(&attest_data.certs)?;
+        send_msg(&mut stream, &cert_chain_der).await?;
+
+        // get & verify server attestation cert chain
+        let server_cert_chain = recv_msg(&mut stream).await?;
+        let server_cert_chain = certs_from_der(&server_cert_chain)?;
+        let root =
+            dice_verifier::verify_cert_chain(&server_cert_chain, Some(&roots))?;
+        let server_platform_id =
+            dice_mfg_msgs::PlatformId::try_from(&server_cert_chain)?;
+        info!(
+            log,
+            "Cert chain from peer \"{}\" verified against root \"{}\"",
+            server_platform_id.as_str()?,
+            root.tbs_certificate.subject,
+        );
+
+        // send measurement log to server
+        let mut buf = vec![0u8; Log::MAX_SIZE];
+        let log_len = hubpack::serialize(&mut buf, &attest_data.log)?;
+        send_msg(&mut stream, &buf[..log_len]).await?;
+
+        // get measurement log from server
+        let server_log = recv_msg(&mut stream).await?;
+        let (server_log, _): (Log, _) = hubpack::deserialize(&server_log)?;
+
+        // hubpack attestation and send to server
+        let mut buf = vec![0u8; Attestation::MAX_SIZE];
+        let len = hubpack::serialize(&mut buf, &attest_data.attestation)?;
+        send_msg(&mut stream, &buf[..len]).await?;
+
+        // get attestation from server
+        let server_attestation = recv_msg(&mut stream).await?;
+        let (server_attestation, _): (Attestation, _) =
+            hubpack::deserialize(&server_attestation)?;
+
+        // verify server attestation
+        dice_verifier::verify_attestation(
+            &server_cert_chain[0],
+            &server_attestation,
+            &server_log,
+            &nonce,
+        )?;
+        info!(log, "Peer attestation verified");
+
+        // appraise measurements from server attestation against reference
+        // measurements
+        let measurements =
+            MeasurementSet::from_artifacts(&server_cert_chain, &server_log)?;
+        dice_verifier::verify_measurements(
+            &measurements,
+            &reference_measurements,
+        )?;
```

Comment:
If we want to actually make measurements non-blocking we need to not use `?` here. I accidentally tested this on racklette where I forgot to update one sled with a newer SP not in the corpus and rack setup failed.